File: MediSentinel_SIH-feature-frontend-risk-alerts/MediSentinel_SIH-feature-frontend-risk-alerts/backend/src/services/differentialPrivacy.js
// Differential Privacy Laplace Mechanism Engine

/**
 * Generates random noise drawn from the Laplace distribution Laplace(0, scale)
 * where scale b = sensitivity / epsilon
 */
export function sampleLaplace(scale) {
  if (scale <= 0) return 0;
  const u = Math.random() - 0.5;
  // Sign(u) * ln(1 - 2|u|)
  const noise = -scale * Math.sign(u) * Math.log(1 - 2 * Math.abs(u));
  return noise;
}

/**
 * Perturbs a raw time series data array using Laplace Differential Privacy
 * @param {Array} data - Array of data points with numeric counts
 * @param {number} epsilon - Privacy loss budget parameter (epsilon > 0)
 * @param {number} sensitivity - Global sensitivity Delta f (default: 1.0)
 */
export function applyDifferentialPrivacy(data = [], epsilon = 1.0, sensitivity = 1.0) {
  const safeEpsilon = Math.max(0.01, parseFloat(epsilon) || 1.0);
  const scale = sensitivity / safeEpsilon;

  const perturbedData = data.map((point) => {
    const rawCount = point.fever !== undefined ? point.fever : (point.value || 0);
    const noise = sampleLaplace(scale);
    const noisyCount = Math.max(0, Math.round(rawCount + noise));

    return {
      ...point,
      rawCount,
      noisyCount,
      noiseApplied: parseFloat(noise.toFixed(2)),
      epsilon: safeEpsilon,
      scale: parseFloat(scale.toFixed(2)),
    };
  });

  return {
    epsilon: safeEpsilon,
    sensitivity,
    scale: parseFloat(scale.toFixed(2)),
    guarantee: `(ε=${safeEpsilon}, δ=0)-Differential Privacy`,
    mathFormula: `Noise ~ Laplace(0, ${sensitivity} / ${safeEpsilon})`,
    data: perturbedData,
  };
}

File: MediSentinel_SIH-feature-frontend-risk-alerts/MediSentinel_SIH-feature-frontend-risk-alerts/backend/app/services/differential_privacy.py
"""Differential Privacy Laplace Mechanism Engine."""
import math
import random
from typing import Any


def sample_laplace(scale: float) -> float:
    if scale <= 0:
        return 0.0
    u = random.random() - 0.5
    sign = 1.0 if u > 0 else (-1.0 if u < 0 else 0.0)
    noise = -scale * sign * math.log(1.0 - 2.0 * abs(u))
    return noise


def apply_differential_privacy(
    data: list[dict[str, Any]],
    epsilon: float = 1.0,
    sensitivity: float = 1.0,
) -> dict[str, Any]:
    safe_epsilon = max(0.01, float(epsilon))
    scale = sensitivity / safe_epsilon

    perturbed_data = []
    for point in data:
        raw_count = point.get("fever", point.get("rawCount", point.get("value", 0)))
        noise = sample_laplace(scale)
        noisy_count = max(0, round(raw_count + noise))

        perturbed_data.append({
            **point,
            "rawCount": raw_count,
            "noisyCount": noisy_count,
            "noiseApplied": round(noise, 2),
            "epsilon": safe_epsilon,
            "scale": round(scale, 2),
        })

    return {
        "epsilon": safe_epsilon,
        "sensitivity": sensitivity,
        "scale": round(scale, 2),
        "guarantee": f"(ε={safe_epsilon}, δ=0)-Differential Privacy",
        "mathFormula": f"Noise ~ Laplace(0, {sensitivity} / {safe_epsilon})",
        "data": perturbed_data,
    }

